sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/WindowFn.java

```diff
@@ -118,6 +118,15 @@ public abstract void merge(Collection<W> toBeMerged, W mergeResult)
   public abstract boolean isCompatible(WindowFn<?, ?> other);
 
   /**
+   * Throw {@link IncompatibleWindowException} if this WindowFn does not perform
+   * the same merging as the given ${@code WindowFn}.
+   *
+   * @throws IncompatibleWindowException if compared WindowFns are not compatible.
+   */
+  public abstract void verifyCompatibility(WindowFn<?, ?> other)
```

Comment:
This may not make it into the first stable release. Then this would be backwards-incompatible. To have compatibility, we would make this call `isCompatible` and issue some default error message. But the javadoc should encourage users to override this method instead.